### ISSUE TITLE
0.10.0.0

### ISSUE THREAD

**Issue by seagreen** [seagreen/hjsonschema#31]
- [x] - Fix link to `./examples/Standard.hs` in README.
- [x] - Build examples as tests, not executables.
- [x] - Add a function to fetch referenced schemas from the filesystem instead of over HTTP.
- [x] - Switch from a POSIX to Perl style regex library.
- [x] - Pick nicer names for some of the errors
- [x] - Get GHC 7.8 working again.
- [x] - Update examples, haddocks, changelog.

**Comment by mdorman** [comment]
The fact that hjsonpointer and hjsonschema both default to creating an executable named 'example' has rendered them impossible to install side-by-side in nixos without hacking the packaging.  I imagine this is true in debian/ubuntu as well.

Perhaps you could set them to not be built by default, or at least provide a flag to turn off adding the executables?

**Comment by seagreen** [comment]
Perhaps the examples should become part of the library, like pipe's [Pipes.Tutorial](https://hackage.haskell.org/package/pipes-4.1.8/docs/Pipes-Tutorial.html)? If I only provide an optional flag to turn them off that's one more thing for users to worry about, and if I provide a flag to turn them _on_ then people (probably me) won't be forced to build them when they build the library, and we'll no longer have a guarantee that the examples actually work.

Or perhaps a better idea, I could write the examples as tests, then we'll have verification that they both compile and the output's what we expect.

Thoughts?

**Comment by mdorman** [comment]
I think if they can be re-cast as tests, that's best of all.  Including them as part of the library doesn't seem unreasonable either.

I will observe that having a flag to turn off the executables is probably not too onerous---the community of people who have to worry about it is probably very, very small.

**Comment by seagreen** [comment]
Thanks for the report. Switched to tests here: 706d67e068f5568e5a73718947eaf41d342128d5.

**Comment by seagreen** [comment]
I'll give this release until mid-next week so we have time to discover any new defects. The current release branch is [here](https://github.com/seagreen/hjsonschema/tree/fetch-improvements).

**Comment by seagreen** [comment]
Final code (including some slight changes to error names since `fetch-improvements`) is on the [fetch-improvements-rebased](https://github.com/seagreen/hjsonschema/tree/fetch-improvements-rebased) branch. I'm planning to only make changes to the examples and haddocks before release.

**Comment by seagreen** [comment]
The latest version of the library is now on this branch: https://github.com/seagreen/hjsonschema/tree/more-0.10-changes

We used to have two failure types that differed only in one being specialized for Draft4:

```
data Failure err = Failure
  { _failureValidatorsCalled :: !err
  , _failureFinalValidator   :: !Value
  , _failureOffendingData    :: !P.Pointer
  } deriving (Eq, Show)

```

and

```
-- | A version of 'FR.Failure' specialized for JSON Schema Draft 4.
data Failure = Failure
  { _failureValidatorsCalled :: !ValidatorChain
  , _failureFinalValidator   :: !Value
  , _failureOffendingData    :: !P.Pointer
  } deriving (Eq, Show)
```

Now we just have one data type:

```
data Failure err = Invalid
  { _invalidValidatorsCalled :: !err
  , _invalidFinalValidator   :: !Value
  , _invalidOffendingData    :: !P.Pointer
  } deriving (Eq, Show)
```

And a type alias to specialize it for Draft 4:

```
type Invalid = FR.Failure ValidatorChain
```

This is way better, the old way was hacky.

**Comment by seagreen** [comment]
Good to go, will push to Hackage soon if no one finds more problems.

**Comment by int-index** [comment]
I can confirm that the latest version (`bd8bd187c26f89eed7a8d5351f08f498aa0382d6`) works for me. You might want to update the README.md and Changelog to mention that now a better regex library is used.

**Comment by seagreen** [comment]
I actually already changed the README (I still need to leave a note there because I'm still not using a ECMAScript style regex library, though moving from a Posix to a Perl style one is closer than I had before). Good call suggesting I update the changelog though!

Final commit [here](https://github.com/seagreen/hjsonschema/commit/1a64de8637582494d38794844359e08b205bc656) and released to Hackage as 0.10.0.0.

**Comment by seagreen** [comment]
PS Thanks for all your excellent help @int-index, it's very appreciated.
